README
You'll find here some pieces of code showing how to use the
GroupCamp public provisioning and reporting APIs.

Manifest:
- group.py shows, based on standard Python libraries, how to create
  a project and add users in that project.
- track.py shows how to fetch time-logs from the GroupCamp
  application



ChangeLog.md
# GroupCamp API Release notes

2019-07-16 - Updates for GroupCamp v3 and new API verbs

* There is no more `managing_team` associated with a group.
* Published the API to fetch informations about the budget of a project.
* Published the API to fetch the report on projects.

2018-09-25 - Managing organizations (Companies)

* Published the API to create, get, edit and list invited organizations (aka
  "Companies" in the GroupCamp web interface).
* `orga` and `subtype` fields can be set when creating/editing a project. When
  the subtype is `customer` (for customer's projects), the `orga` field allows
  to choose the customer's company.
* Example of code has been added to the `orga.py` script.

2018-09-25 - Tag management for projects
* Fields `rich_tags` (tag's uuid, name, color and where it may be used) and
  `tag_names` (only the names) are added to the informations returned to
  describe a project.
* The field `tag_names` can be used to change/set the tags of a project when
  creating or editing.
* New `STag` type definition
* Example of tag management code has been added to the `group.py` script.



orga.py
from hashlib import sha1
import hmac
import httplib
import json
import pprint
import re

pp = pprint.PrettyPrinter(indent=4)

key =  "Get it from the admin panel, in the API icon"
secret = "Get it from the admin panel, in the API icon"
debug = 1
acting = 'big.boss@localhost'


def do_call (method, path, body, q_string='') :

	# The text to be signed is a concatenation of
	# - the method (POST, PUT, GET, etc),
	# - the string '::'
	# - the path, with the leading /, without the trailing /
	# - the string '::'
	# - the body of the request (can be a json object, or an empty string)

	text = method + '::' + path + '::' + body
	signature = hmac.new(secret, text, sha1)
	sig = signature.digest().encode("base64").rstrip('\n')

	# X-Gcmp-Application MUST be 'provisioning-1'
	# X-Gcmp-Acting MUST be set to the user which will be loged as performing the action.
	#   You can create a "fake" user in your GroupCamp account for this purpose. Her
	#   privileges are not checked - this user is *only* loged as the user who did
	#   perform the action.
	# Authorization is the string "GCMP <key>:<signature>", the secret MUST NEVER BE SENT.
	#   It should be kept secret.
	# Content-Type is application/json, since the body is a json object.

	headers = {
		'X-Gcmp-Application': 'provisioning-1',
		'X-Gcmp-Acting': acting,
		'Authorization': 'GCMP '+key+':'+sig,
		'Content-Type': 'application/json'
	}

	if debug :
		print "The text to be signed <"+text+'>'
		print "The obtained signature "+sig

	if method == 'GET' and q_string != '':
		path = path + '?' + q_string

	cnx = httplib.HTTPSConnection('api.mustrum.groupcamp.test', 443)
	cnx.request(method, path, body, headers)
	res = cnx.getresponse()
	return res

def do_call_url ( uri ) :
	res = re.match(r"^(https?)://([^/]+)([^?]+)?(.*)$",uri)
	if not res :
		print "No match"
		return None
	protocol = res.group(1)
	domain   = res.group(2)
	path     = res.group(3)
	args     = res.group(4)
	print "Protocol ", protocol
	print "Domain   ", domain
	print "Path     ", path
	print "Q-string ", args
	return do_call('GET', path, '', args)




def new_orga(name, description):
	data = {
		'name' : name,
		'description' : description,
	}
	body = json.dumps(data)

	print body

	res = do_call('POST', '/core/v1/orga/invited', body)
	status = res.status
	body = res.read()
	data = json.loads(body)

	if debug :
		print "Status ",status
		print "Raw answer ",body

	if status != 200 :
		print "Failed"
		if data['error'] == 'bad_request' :
			print "Bad request";
			if data['errors'][0]['field'] == 'name' and data['errors'][0]['error'] == 'already_use' :
				print "Duplicate name"
		return 'failed'
	else :
		print "Creation successful, orga ID is " + data['result']['id']
		return data['result']['id']

def change_orga(uuid, name, description):
	data = {
		'name' : name,
		'description' : description,
	}
	body = json.dumps(data)

	res = do_call('POST', '/core/v1/orga/invited/'+uuid, body)
	status = res.status
	body = res.read()
	data = json.loads(body)

	if debug :
		print "Status ",status
		print "Raw answer ",body

	if status != 200 :
		print "Failed"
		if data['error'] == 'bad_request' :
			print "Bad request"
			if data['errors'][0]['field'] == 'name' and data['errors'][0]['error'] == 'already_use' :
				print "Duplicate name in orga modification"
		return 'failed';
	else :
		print "Modification successful, organization ID is " + data['result']['id']
		return data['result']['id']



def get_orga(uuid) :
	res = do_call('GET', '/core/v1/orga/'+uuid, '', '')
	body = res.read()
	data = json.loads(body)
	pp.pprint(data)
	return data['result']

def get_orgas() :
	res = do_call('GET', '/core/v1/orgas/invited', '', '')
	body = res.read()
	data = json.loads(body)
	pp.pprint(data)


new_uuid = new_orga('API 1', 'A company created via API')
change_orga(new_uuid, 'API 1 renamed', 'New description for the organization')
get_orga(new_uuid)

get_orgas()


task.js
/*
	This tiny test script is designed to be used in a browser.

	Include those in the HEAD of your HTML page:
	<script src="https://cdnjs.cloudflare.com/ajax/libs/crypto-js/3.1.9-1/crypto-js.min.js"></script>
	<script src="https://cdnjs.cloudflare.com/ajax/libs/crypto-js/3.1.9-1/hmac-sha1.min.js"></script>
	<script src="https://cdnjs.cloudflare.com/ajax/libs/crypto-js/3.1.9-1/enc-base64.min.js"></script>
 */

// This one is the key, declared for your API access in the GroupCamp admin interface.
var key    = 'The key';
// This one is the secret associated with the key, used to sign your request. Must not
// be sent to GroupCamp (or elsewhere).
var secret = 'The secret associated to the key';
// The main GroupCamp API server
var domain = 'api.groupcamp.com';
// The user acting
var acting = 'e-mail address of the acting user';

/*
	Will send a request to the GroupCamp API, and display the result in the web page.
*/
function send_request(method, path, body, q_string, on_success, on_error) {
	// Signing request
	var text = method + '::' + path + '::' + body
	var hash = CryptoJS.HmacSHA1(text, secret);
	var hashInBase64 = CryptoJS.enc.Base64.stringify(hash);

	// Sending request
	var url = 'https://'+domain+path;
	if ( q_string ) {
		url += '?'+q_string; 
	} 
	var xhr = new XMLHttpRequest();
	xhr.open(method, url, true);
	xhr.onload = function(res) { on_success(xhr.response); };
	xhr.onerror = on_error;
	xhr.withCredentials = false;
	xhr.responseType = 'json';
	xhr.setRequestHeader('Content-Type','application/json');
	xhr.setRequestHeader('X-Gcmp-Application','reporting-1');
	xhr.setRequestHeader('X-Gcmp-Acting',acting);
	xhr.setRequestHeader('Authorization','GCMP '+key+':'+hashInBase64);
	xhr.send(body);
}

function get_task(uuid, on_success, on_error) {
	var method = 'GET';
	var path = '/task/v1/task/' + uuid;
	var body = '';
	var q_string = ''
	send_request(method, path, body, q_string, on_success, on_error);
}

var uuid = 'Uuid of a task (you can get it from the information panel)';
get_task(uuid, function(res) {console.info(res);}, function(res) {console.error(res);});



task.py
from hashlib import sha1
import hmac
import httplib
import json
import pprint
import re

pp = pprint.PrettyPrinter(indent=4)

host   = 'api.groupcamp.com'
key    = "token from the API icon in the admin panel of your GroupCamp account"
secret = "the secret associated with the above key"
debug  = 1
acting = 'email address of the "acting" user (not that important for the reporting API)'


def do_call (method, path, body='', q_string='') :

	# The text to be signed is a concatenation of
	# - the method (POST, PUT, GET, etc),
	# - the string '::'
	# - the path, with the leading /, without the trailing /
	# - the string '::'
	# - the body of the request (can be a json object, or an empty string)

	text = method + '::' + path + '::' + body
	signature = hmac.new(secret, text, sha1)
	sig = signature.digest().encode("base64").rstrip('\n')

	# X-Gcmp-Application MUST be 'reporting-1'
	# X-Gcmp-Acting MUST be set to the user which will be logged as performing the action.
	#   You can create a "fake" user in your GroupCamp account for this purpose. User
	#   privileges are not checked - this user is *only* logged as the user who did
	#   perform the action.
	# Authorization is the string "GCMP <key>:<signature>", the secret MUST NEVER BE SENT.
	#   It should be kept secret.
	# Content-Type is application/json, since the body is a json object.

	headers = {
		'X-Gcmp-Application': 'reporting-1',
		'X-Gcmp-Acting': acting,
		'Authorization': 'GCMP '+key+':'+sig,
		'Content-Type': 'application/json'
	}

	if debug :
		print "The text to be signed <"+text+'>'
		print "The obtained signature "+sig

	if method == 'GET' and q_string != '':
		path = path + '?' + q_string

	cnx = httplib.HTTPSConnection(host, 443)
	cnx.request(method, path, body, headers)
	res = cnx.getresponse()
	return res



def get_element(method, path) :
	res = do_call(method, path)
	status = res.status
	body = res.read()

	if debug :
		print "Status ",status
		print "Raw answer ",body

	data = json.loads(body)

	if status != 200 :
		print "Failed"
		return 'failed'
	else :
		print "Request successful"
		result = data['result']
	return result


result = get_element('GET', '/task/v1/task/[UUID of a task]')
pp.pprint(result)


track.py
from hashlib import sha1
import hmac
import httplib
import json
import pprint
import re

pp = pprint.PrettyPrinter(indent=4)

key    = "Get a key, for interface 'reporting', from the admin panel, in the API icon"
secret = "Get the secret associated with the key, in the same place"
debug  = 1
acting = 'big.boss@localhost'


def do_call (method, path, body, q_string='') :

	# The text to be signed is a concatenation of
	# - the method (POST, PUT, GET, etc),
	# - the string '::'
	# - the path, with the leading /, without the trailing /
	# - the string '::'
	# - the body of the request (can be a json object, or an empty string)

	text = method + '::' + path + '::' + body
	signature = hmac.new(secret, text, sha1)
	sig = signature.digest().encode("base64").rstrip('\n')

	# X-Gcmp-Application MUST be 'reporting-1'
	# X-Gcmp-Acting MUST be set to the user which will be logged as performing the action.
	#   You can create a "fake" user in your GroupCamp account for this purpose. User
	#   privileges are not checked - this user is *only* logged as the user who did
	#   perform the action.
	# Authorization is the string "GCMP <key>:<signature>", the secret MUST NEVER BE SENT.
	#   It should be kept secret.
	# Content-Type is application/json, since the body is a json object.

	headers = {
		'X-Gcmp-Application': 'reporting-1',
		'X-Gcmp-Acting': acting,
		'Authorization': 'GCMP '+key+':'+sig,
		'Content-Type': 'application/json'
	}

	if debug :
		print "The text to be signed <"+text+'>'
		print "The obtained signature "+sig

	if method == 'GET' and q_string != '':
		path = path + '?' + q_string

	cnx = httplib.HTTPSConnection('api.groupcamp.com', 443)
	cnx.request(method, path, body, headers)
	res = cnx.getresponse()
	return res

def do_call_url ( uri ) :
	# When an API is paginated, the next_page and previous_page provides
	# links to the sibbling pages.
	# The python httplib library needs to split this URI into various
	# pieces (host, path, query string, etc). So we de rely on the same
	# do_call as before, since we need to split the URI anyway. Can probably
	# be written in a better way in your usual framework :)
	res = re.match(r"^(https?)://([^/]+)([^?]+)\?(.*)$",uri)
	if not res :
		print "No match"
		return None
	protocol = res.group(1)
	domain   = res.group(2)
	path     = res.group(3)
	args     = res.group(4)
	print "Protocol ", protocol
	print "Domain   ", domain
	print "Path     ", path
	print "Q-string ", args

	# Protocol will always be https. Domain will always be the same as the domain
	# of the first request.
	return do_call('GET', path, '', args)


def get_all_lines(method, path, body, query_string) :
	res = do_call(method, path, body, query_string)
	status = res.status
	body = res.read()

	if debug :
		print "Status ",status
		print "Raw answer ",body

	data = json.loads(body)

	if status != 200 :
		print "Failed"
		if data['error'] == 'bad_request' :
			print "Bad request";
			if data['errors'][0]['field'] == 'name' and data['errors'][0]['error'] == 'already_use' :
				print "Duplicate name"
		return 'failed'
	else :
		print "Request successful, number of pages:", data["page_count"]
		result = data['result']

	count = data['page_count']
	print "There are", count, "pages of result"
	# pp.pprint(data)
	while 'next_page' in data :
		res = do_call_url(data['next_page'])
		if not res :
			return 'failed on a page'
		body = res.read()
		data = json.loads(body)
		result.extend(data['result'])
		# pp.pprint(data)

	print "Obtained", len(result), "records :"
	return result

def get_tlogs(start_date, end_date):
        res = do_call('GET', '/track/v1/timelogs', '', 'start='+start_date+'&end='+end_date)
        body = res.read();
        data = json.loads(body)
        pp.pprint(data)

# Get logs for precisely one week
get_tlogs('2016-10-24','2016-10-30')




budget.py
from hashlib import sha1
import hmac
import httplib
import json
import pprint
import re

pp = pprint.PrettyPrinter(indent=4)

key    =  "Get it from the admin panel, in the API icon"
secret = "Get it from the admin panel, in the API icon"
debug  = 1
acting = 'big.boss@localhost'


def do_call (method, path, body, q_string='') :

	# The text to be signed is a concatenation of
	# - the method (POST, PUT, GET, etc),
	# - the string '::'
	# - the path, with the leading /, without the trailing /
	# - the string '::'
	# - the body of the request (can be a json object, or an empty string)

	text = method + '::' + path + '::' + body
	signature = hmac.new(secret, text, sha1)
	sig = signature.digest().encode("base64").rstrip('\n')

	# X-Gcmp-Application MUST be 'reporting-1'
	# X-Gcmp-Acting MUST be set to the user which will be loged as performing the action.
	#   You can create a "fake" user in your GroupCamp account for this purpose. Her
	#   privileges are not checked - this user is *only* loged as the user who did
	#   perform the action.
	# Authorization is the string "GCMP <key>:<signature>", the secret MUST NEVER BE SENT.
	#   It should be kept secret.
	# Content-Type is application/json, since the body is a json object.

	headers = {
		'X-Gcmp-Application': 'reporting-1',
		'X-Gcmp-Acting': acting,
		'Authorization': 'GCMP '+key+':'+sig,
		'Content-Type': 'application/json'
	}

	if debug :
		print "The text to be signed <"+text+'>'
		print "The obtained signature "+sig

	if method == 'GET' and q_string != '':
		path = path + '?' + q_string

	cnx = httplib.HTTPSConnection('api.groupcamp.com', 443)
	cnx.request(method, path, body, headers)
	res = cnx.getresponse()
	return res

def get_budget(uuid) :
	res = do_call('GET', '/track/v1/budget/'+uuid, '')
	status = res.status
	body = res.read()

	if debug :
		print "Status ",status
		print "Raw answer ",body

	data = json.loads(body)
	pp.pprint(data)

get_budget('The group UUID found in the information panel of the project')


report.py
from hashlib import sha1
import hmac
import httplib
import json
import pprint
import re

pp = pprint.PrettyPrinter(indent=4)

key    = "Get it from the admin panel, in the API icon"
secret = "Get it from the admin panel, in the API icon"
debug  = 1
acting = 'big.boss@localhost'


def do_call (method, path, body, q_string='') :

	# The text to be signed is a concatenation of
	# - the method (POST, PUT, GET, etc),
	# - the string '::'
	# - the path, with the leading /, without the trailing /
	# - the string '::'
	# - the body of the request (can be a json object, or an empty string)

	text = method + '::' + path + '::' + body
	signature = hmac.new(secret, text, sha1)
	sig = signature.digest().encode("base64").rstrip('\n')

	# X-Gcmp-Application MUST be 'reporting-1'
	# X-Gcmp-Acting MUST be set to the user which will be loged as performing the action.
	#   You can create a "fake" user in your GroupCamp account for this purpose. Her
	#   privileges are not checked - this user is *only* loged as the user who did
	#   perform the action.
	# Authorization is the string "GCMP <key>:<signature>", the secret MUST NEVER BE SENT.
	#   It should be kept secret.
	# Content-Type is application/json, since the body is a json object.

	headers = {
		'X-Gcmp-Application': 'reporting-1',
		'X-Gcmp-Acting': acting,
		'Authorization': 'GCMP '+key+':'+sig,
		'Content-Type': 'application/json'
	}

	if debug :
		print "The text to be signed <"+text+'>'
		print "The obtained signature "+sig

	if method == 'GET' and q_string != '':
		path = path + '?' + q_string

	cnx = httplib.HTTPSConnection('api.groupcamp.com', 443)
	cnx.request(method, path, body, headers)
	res = cnx.getresponse()
	return res


def get_report() :
	res = do_call('GET', '/track/v1/report/', '', 'search=Some%20text')
	status = res.status
	body = res.read()

	if debug :
		print "Status ",status
		print "Raw answer ",body

	data = json.loads(body)
	pp.pprint(data)

get_report()


group.py
from hashlib import sha1
import hmac
import httplib
import json
import pprint

pp = pprint.PrettyPrinter(indent=4)

key =  "Get it from the admin panel, in the API icon"
secret = "Get it from the admin panel, in the API icon"
debug = 1
acting = 'big.boss@localhost'

def do_call (method, path, body, q_string='') :

	# The text to be signed is a concatenation of
	# - the method (POST, PUT, GET, etc),
	# - the string '::'
	# - the path, with the leading /, without the trailing /
	# - the string '::'
	# - the body of the request (can be a json object, or an empty string)

	text = method + '::' + path + '::' + body
	signature = hmac.new(secret, text, sha1)
	sig = signature.digest().encode("base64").rstrip('\n')

	# X-Gcmp-Application MUST be 'provisioning-1'
	# X-Gcmp-Acting MUST be set to the user which will be logged as performing the action.
	#   You can create a "fake" user in your GroupCamp account for this purpose. User
	#   privileges are not checked - this user is *only* logged as the user who did
	#   perform the action.
	# Authorization is the string "GCMP <key>:<signature>", the secret MUST NEVER BE SENT.
	#   It should be kept secret.
	# Content-Type is application/json, since the body is a json object.

	headers = {
		'X-Gcmp-Application': 'provisioning-1',
		'X-Gcmp-Acting': acting,
		'Authorization': 'GCMP '+key+':'+sig,
		'Content-Type': 'application/json'
	}

	if debug :
		print "The text to be signed <"+text+'>'
		print "The obtained signature "+sig

	if method == 'GET' and q_string != '':
		path = path + '?' + q_string

	cnx = httplib.HTTPSConnection('api.groupcamp.com', 443)
	cnx.request(method, path, body, headers)
	res = cnx.getresponse()
	return res





def new_project(name, description, access, guests, gcat_name, management_team):
	data = {
		'name' : name,
		'description' : description,
		'access' : access,
		'with_guests' : guests,
		'category_name' : gcat_name,
#		'leader1' : 'UUID or e-mail',
#		'leader2' : 'UUID or e-mail',
#		'managers' : [ 'UUID or e-mail', ... ],
#		'template' : {
#			'id'      : 'UUID of an existing project',
#			'track'   : 1,   # Will copy the budget informations
#			'members' : 1,   # Will copy the members
#			'topics'  : 1,   # Will copy the topics from 'Discussion' application
#			'files' : {
#				'folders' : 1, # Will copy the foldes-tree
#				'owned'   : 1, # Will copy the files owned by the project (required 'folders')
#				'others'  : 1, # Will copy the other files
#			}
#		},
#		'subtype' : 'customer', # or 'internal'
#		'orga' : '<the_uuid_of_the_orga>', # only for 'customer' projects
		'management_team' : team_uuid
	}
	body = json.dumps(data)

	res = do_call('POST', '/core/v1/group/project', body)
	status = res.status
	body = res.read()
	data = json.loads(body)

	if debug :
		print "Status ",status
		print "Raw answer ",body

	if status != 200 :
		print "Failed"
		if data['error'] == 'bad_request' :
			print "Bad request";
			if data['errors'][0]['field'] == 'name' and data['errors'][0]['error'] == 'already_use' :
				print "Duplicate name"
		return 'failed'
	else :
		print "Creation successful, groupe ID is " + data['result']['id']
		return data['result']['id']

def change_project(uuid, name, description, access, guests):
	data = {
		'name' : name,
		'description' : description,
		'access' : access,
		'with_guests' : guests
	}
	body = json.dumps(data)

	res = do_call('POST', '/core/v1/group/project/'+uuid, body)
	status = res.status
	body = res.read()
	data = json.loads(body)

	if debug :
		print "Status ",status
		print "Raw answer ",body

	if status != 200 :
		print "Failed"
		if data['error'] == 'bad_request' :
			print "Bad request"
			if data['errors'][0]['field'] == 'name' and data['errors'][0]['error'] == 'already_use' :
				print "Duplicate name in project modification"
		return 'failed';
	else :
		print "Modification successful, groupe ID is " + data['result']['id']
		return data['result']['id']

def change_projet_tags(uuid, name, tags):
	data = {
		'name' : name,
		'tags_names': tags
	}
	body = json.dumps(data)
	pp.pprint(data)

	res = do_call('POST', '/core/v1/group/project/'+uuid, body)
	status = res.status
	body = res.read()
	data = json.loads(body)

	if debug :
		print "Status ",status
		print "Raw answer ",body
		pp.pprint(data)

	if status != 200 :
		print "Failed"
		if data['error'] == 'bad_request' :
			print "Bad request"
			if data['errors'][0]['field'] == 'name' and data['errors'][0]['error'] == 'already_use' :
				print "Duplicate name in project modification"
		return 'failed';
	else :
		print "Modification successful, groupe ID is " + data['result']['id']
		return data['result']['id']

def insert_user(group_id, email) :
	res = do_call('PUT', '/core/v1/group/'+group_id+'/user/'+email, '')
	status = res.status
	if status != 200 :
		print "Adding member failed"
		print "Raw answer ", res.read()
	else :
		print "Added member in the group"


def remove_user(group_id, email) :
	res = do_call('DELETE', '/core/v1/group/'+group_id+'/user/'+email, '')
	status = res.status
	if status != 200 :
		print "Removing member failed";
		print "Raw answer ", res.read()
	else :
		print "Removed member from the group"


def user_teams(email) :
	res = do_call('GET', '/core/v1/user/'+email+'/teams', '')
	if res.status != 200 :
		print "Failed"
		return 'failed'

	body = res.read()
	data = json.loads(body)
	for team in data['result'] :
		if team['is_tech'] :
			management = " is a management team"
		else :
			management = ""
		print "Team '"+team['name']+"' having UUID = "+team['id']+management
	return data


team_uuid = 'Get it from the GroupCamp user interface'
gcat_name = 'Main category'
user_email = 'jean@localhost'
group_id = new_project('The new group 3', 'This group have been created via the provisioning API', 'invite', False, gcat_name, team_uuid)
if group_id != 'failed' :
	insert_user(group_id, user_email)
	change_project(group_id, 'Another name', 'The description is modified', 'open', True)
	change_project_tags(group_id, 'Another name', [ 'Tag one', 'Test tag' ])

print "The user '"+user_email+"' is a member of the following teams:"
user_teams(user_email)


